### PR TITLE
Facebook Blob

### PR DESCRIPTION
**Pull request by allentsai**
This commit:
* Adds a FacebookBlob class to the library.
* Adds a Utility trait for convenience.

### CODE REVIEW

File: src/PaperG/FirehoundBlob/Facebook/FacebookBlob.php [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                    ? $this->geographicTargeting->toArray() : null,
+            self::DEMOGRAPHIC_TARGETING => isset($this->demographicTargeting)
+                    ? $this->demographicTargeting->toArray() : null,
+            self::AUDIENCE_TARGETING    => isset($this->audienceTargeting)
+                    ? $this->audienceTargeting->toArray() : null,
+            self::OBJECTS_TO_UPDATE     => $this->objectsToUpdate,
+            self::AD_ACCOUNT_ID         => $this->adAccountId,
+            self::PAGE_ID               => $this->pageId,
+            self::ACCESS_TOKEN          => $this->accessToken,
+            self::AD_SETS               => $this->adSets,
```
Comment (zhiachong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookBlob.php]: Hmm setting AD_SETS twice?

File: src/PaperG/FirehoundBlob/Facebook/FacebookBlob.php [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        ];
+        $adSets = null;
+        if (!empty($this->adSets)) {
+            $adSets = [];
+            foreach($this->adSets as $adSet) {
+                $adSets[] = $adSet->toAssociativeArray();
+            }
+        }
+
+        $array[self::AD_SETS] = $adSets;
```
Comment (zhiachong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookBlob.php]: Setting it again here

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreative.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            self::TYPE => $this->type,
+            self::VERSION => self::CURRENT_VERSION
+        ];
+
+        $objects = [];
+        foreach ($this->objects as $creativeData) {
+            $objects[] = $creativeData->toArray();
+        }
+
+        $array[self::OBJECTS] = $objects;
```
Comment (zhiachong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreative.php]: I was thinking about this. Should we just not set it if it's empty?

File: src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookDemographicTargeting.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace PaperG\FirehoundBlob\Facebook\Targeting;
+
+
+use PaperG\FirehoundBlob\BlobInterface;
+use PaperG\FirehoundBlob\Utility;
+
+class FacebookDemographicTargeting implements  BlobInterface
```
Comment (zhiachong) [inline src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookDemographicTargeting.php]: There's an extra space

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    private $description = null;
+    private $variationId = null;
+
+    public function __construct($array = null)
+    {
+        $this->fromArray($array);
+    }
+
+    /**
+     * @param null $callToAction
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: This is a string right?

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    /**
+     * @return null
+     */
+    public function getCaption()
+    {
+        return $this->caption;
+    }
+
+    /**
+     * @param null $description
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: String?

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    /**
+     * @return null
+     */
+    public function getCallToAction()
+    {
+        return $this->callToAction;
+    }
+
+    /**
+     * @param null $caption
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: String?

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    /**
+     * @return null
+     */
+    public function getDescription()
+    {
+        return $this->description;
+    }
+
+    /**
+     * @param null $landingPage
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: String?

File: src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    private $type;
+    private $ids;
+
+    public function __construct($array = null)
+    {
+        $this->fromArray($array);
+    }
+
+    /**
+     * @param mixed $ids
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php]: int[] ?

File: src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * @return mixed
+     */
+    public function getIds()
+    {
+        return $this->ids;
+    }
+
+    /**
+     * @param mixed $type
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php]: int[]?

File: src/PaperG/FirehoundBlob/BlobInterface.php [hunk 11]
```diff
@@ -13,6 +13,4 @@ public function toArray();
      * @param $array array
      */
     public function fromArray($array);
-
-    public function validate();
```
Comment (a37wong) [inline src/PaperG/FirehoundBlob/BlobInterface.php]: What was the conclusion about the validate method?  I notice FacebookBlob lacks any validate method.

Comment (allentsai) [inline src/PaperG/FirehoundBlob/BlobInterface.php]: Aiming on using validators, so they won't be on the blob.  I'm pulling the BasicInfo validation off when we write the validators.

File: src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php (hunk 10 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php]: stupid auto complete

File: src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php (hunk 9 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/Targeting/FacebookAudienceTargeting.php]: stupid auto complete

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php (hunk 8 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: stupid auto complete

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php (hunk 7 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: stupid auto complete

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php (hunk 6 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: stupid auto complete

File: src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php (hunk 5 above)
Comment (allentsai) [inline src/PaperG/FirehoundBlob/Facebook/FacebookCreativeData.php]: stupid auto complete

**Pull request closed without merging**